File: bottlenote-product-api/src/test/java/app/global/security/AccessControlSecurityIntegrationTest.java
package app.global.security;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.http.HttpStatus.FORBIDDEN;
import static org.springframework.http.HttpStatus.TOO_MANY_REQUESTS;

import app.bottlenote.IntegrationTestSupport;
import app.bottlenote.global.security.accesscontrol.AccessControlService;
import app.bottlenote.global.security.accesscontrol.AccessControlStore;
import app.bottlenote.global.security.accesscontrol.AccessControlStore.UnavailableException;
import app.bottlenote.global.security.accesscontrol.BanSnapshotRefresher;
import app.bottlenote.global.security.accesscontrol.RedisAccessControlStore;
import java.time.Duration;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;
import org.springframework.context.annotation.Primary;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.TestPropertySource;

/**
 * access-control을 이 클래스에서만 활성화하고, 실제 Redis + Security filter chain을 검증한다. application-test.yml
 * 기본값(enabled=false)은 다른 통합 테스트에 유지된다.
 */
@Tag("integration")
@DisplayName("[integration] Product AccessControl Security chain")
@ActiveProfiles({"test", "access-control-it"})
@TestPropertySource(properties = "bottlenote.access-control.burst-admission.cooldown=0s")
@Import(AccessControlSecurityIntegrationTest.AccessControlStoreTestConfiguration.class)
class AccessControlSecurityIntegrationTest extends IntegrationTestSupport {

  private static final String PUBLIC_PATH = "/api/v1/alcohols/search";

  @Autowired private AccessControlService accessControlService;
  @Autowired private StringRedisTemplate stringRedisTemplate;
  @Autowired private ControllableAccessControlStore accessControlStore;
  @Autowired private BanSnapshotRefresher banSnapshotRefresher;

  @BeforeEach
  void cleanAccessControlRedisState() {
    Set<String> keys = stringRedisTemplate.keys("bn:ac:*");
    if (keys != null && !keys.isEmpty()) {
      stringRedisTemplate.delete(keys);
    }
    accessControlStore.reset();
    banSnapshotRefresher.refresh();
  }

  @Test
  @DisplayName("허용 IP 요청은 Security chain을 통과한다")
  void 허용_IP_요청_통과() {
    String clientIp = nextTestIp();

    var result =
        mockMvcTester.get().uri(PUBLIC_PATH).header("X-Forwarded-For", clientIp).exchange();

    result.assertThat().hasStatusOk();
    assertThat(result.getResponse().getHeader("X-RateLimit-Limit")).isEqualTo("3");
  }

  @Test
  @DisplayName("ban된 IP 요청은 403을 반환한다")
  void ban된_IP_요청_403() {
    String clientIp = nextTestIp();
    accessControlService.banIp(clientIp, Duration.ofMinutes(5), "integration-ban");

    var result =
        mockMvcTester.get().uri(PUBLIC_PATH).header("X-Forwarded-For", clientIp).exchange();

    result.assertThat().hasStatus(FORBIDDEN);
    assertThat(result.getResponse().getHeader("Retry-After")).isNotBlank();
  }

  @Test
  @DisplayName("default rate limit을 초과하면 429를 반환한다")
  void rate_limit_초과_429() {
    String clientIp = nextTestIp();

    for (int i = 0; i < 3; i++) {
      mockMvcTester
          .get()
          .uri(PUBLIC_PATH)
          .header("X-Forwarded-For", clientIp)
          .exchange()
          .assertThat()
          .hasStatusOk();
    }

    var limited =
        mockMvcTester.get().uri(PUBLIC_PATH).header("X-Forwarded-For", clientIp).exchange();

    limited.assertThat().hasStatus(TOO_MANY_REQUESTS);
    assertThat(limited.getResponse().getHeader("Retry-After")).isNotBlank();
    assertThat(limited.getResponse().getHeader("X-RateLimit-Remaining")).isEqualTo("0");
  }

  @Test
  @DisplayName("Redis 장애에도 fresh snapshot의 ban은 403을 반환한다")
  void Redis_장애_snapshot_ban_403() {
    String clientIp = nextTestIp();
    accessControlService.banIp(clientIp, Duration.ofMinutes(5), "snapshot-ban");
    banSnapshotRefresher.refresh();
    accessControlStore.failRequests();

    var result =
        mockMvcTester.get().uri(PUBLIC_PATH).header("X-Forwarded-For", clientIp).exchange();

    result.assertThat().hasStatus(FORBIDDEN);
    assertThat(accessControlStore.getTryConsumeCalls()).isZero();
  }

  @Test
  @DisplayName("Redis 복구 후 정상 allow와 rate limit 헤더를 반환한다")
  void Redis_복구_정상_allow_및_rate_limit_헤더() {
    String clientIp = nextTestIp();
    accessControlStore.failRequests();

    mockMvcTester
        .get()
        .uri(PUBLIC_PATH)
        .header("X-Forwarded-For", clientIp)
        .exchange()
        .assertThat()
        .hasStatusOk();
    assertThat(accessControlStore.getTryConsumeCalls()).isZero();

    accessControlStore.recover();
    var recovered =
        mockMvcTester.get().uri(PUBLIC_PATH).header("X-Forwarded-For", clientIp).exchange();

    recovered.assertThat().hasStatusOk();
    assertThat(recovered.getResponse().getHeader("X-RateLimit-Limit")).isEqualTo("3");
    assertThat(recovered.getResponse().getHeader("X-RateLimit-Remaining")).isEqualTo("2");
    assertThat(accessControlStore.getTryConsumeCalls()).isEqualTo(1);
  }

  /** 문서용 TEST-NET 대역에서 테스트마다 고유 IP를 생성한다. */
  private static String nextTestIp() {
    int host = Math.floorMod(UUID.randomUUID().hashCode(), 250) + 1;
    return "203.0.113." + host;
  }

  @TestConfiguration(proxyBeanMethods = false)
  static class AccessControlStoreTestConfiguration {

    @Bean
    @Primary
    ControllableAccessControlStore accessControlStore(StringRedisTemplate stringRedisTemplate) {
      return new ControllableAccessControlStore(new RedisAccessControlStore(stringRedisTemplate));
    }
  }

  static class ControllableAccessControlStore implements AccessControlStore {

    private final RedisAccessControlStore delegate;
    private final AtomicBoolean unavailable = new AtomicBoolean(false);
    private final AtomicInteger tryConsumeCalls = new AtomicInteger();

    ControllableAccessControlStore(RedisAccessControlStore delegate) {
      this.delegate = delegate;
    }

    void reset() {
      unavailable.set(false);
      tryConsumeCalls.set(0);
    }

    void failRequests() {
      unavailable.set(true);
    }

    void recover() {
      unavailable.set(false);
    }

    int getTryConsumeCalls() {
      return tryConsumeCalls.get();
    }

    @Override
    public BanLookup lookupBan(String ip) {
      failWhenUnavailable();
      return delegate.lookupBan(ip);
    }

    @Override
    public boolean isBanned(String ip) {
      return delegate.isBanned(ip);
    }

    @Override
    public void ban(String ip, Duration ttl, String reason) {
      delegate.ban(ip, ttl, reason);
    }

    @Override
    public void unban(String ip) {
      delegate.unban(ip);
    }

    @Override
    public void projectBan(String ip, Duration ttl, String reason, long eventId) {
      delegate.projectBan(ip, ttl, reason, eventId);
    }

    @Override
    public void projectUnban(String ip, long eventId) {
      delegate.projectUnban(ip, eventId);
    }

    @Override
    public void removeUnversionedBan(String ip) {
      delegate.removeUnversionedBan(ip);
    }

    @Override
    public BanInfo getBan(String ip) {
      return delegate.getBan(ip);
    }

    @Override
    public List<BanInfo> listBans(int max) {
      failWhenUnavailable();
      return delegate.listBans(max);
    }

    @Override
    public List<BanInfo> listUnversionedBans(int max) {
      return delegate.listUnversionedBans(max);
    }

    @Override
    public ConsumeResult tryConsume(String key, int limit, Duration window) {
      tryConsumeCalls.incrementAndGet();
      failWhenUnavailable();
      return delegate.tryConsume(key, limit, window);
    }

    private void failWhenUnavailable() {
      if (unavailable.get()) {
        throw new UnavailableException(new IllegalStateException("test-controlled Redis outage"));
      }
    }
  }
}

File: bottlenote-admin-api/src/test/kotlin/app/global/security/AccessControlSecurityIntegrationTest.kt
package app.global.security

import app.IntegrationTestSupport
import app.bottlenote.global.security.accesscontrol.AccessControlService
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.test.context.ActiveProfiles
import java.time.Duration
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger

/**
 * access-control을 이 클래스에서만 활성화하고, 실제 Redis + Security filter chain을 검증한다.
 * application-test.yml 기본값(enabled=false)은 다른 통합 테스트에 유지된다.
 */
@Tag("admin_integration")
@DisplayName("[integration] Admin AccessControl Security chain")
@ActiveProfiles("test", "access-control-it")
class AccessControlSecurityIntegrationTest : IntegrationTestSupport() {
	companion object {
		private const val PROTECTED_PATH = "/v1/users"
		private const val BAN_API = "/v1/access-control/ip-bans"
		private val IP_SEQ = AtomicInteger(0)
	}

	@Autowired
	private lateinit var accessControlService: AccessControlService

	private lateinit var accessToken: String

	@BeforeEach
	fun setUpAdminToken() {
		val admin = adminUserTestFactory.persistRootAdmin()
		accessToken = getAccessToken(admin)
	}

	@Test
	@DisplayName("허용 IP 요청은 Security chain을 통과한다")
	fun allowedIpPassesFilterChain() {
		val clientIp = nextTestIp()

		val result =
			mockMvcTester
				.get()
				.uri(PROTECTED_PATH)
				.header("Authorization", "Bearer $accessToken")
				.header("X-Forwarded-For", clientIp)
				.exchange()

		result.assertThat().hasStatusOk()
		assertThat(result.response.getHeader("X-RateLimit-Limit")).isEqualTo("3")
	}

	@Test
	@DisplayName("ban된 IP 요청은 403을 반환한다")
	fun bannedIpReturns403() {
		val clientIp = nextTestIp()
		accessControlService.banIp(clientIp, Duration.ofMinutes(5), "integration-ban")
		try {
			val result =
				mockMvcTester
					.get()
					.uri(PROTECTED_PATH)
					.header("Authorization", "Bearer $accessToken")
					.header("X-Forwarded-For", clientIp)
					.exchange()

			result.assertThat().hasStatus(HttpStatus.FORBIDDEN)
			assertThat(result.response.getHeader("Retry-After")).isNotBlank()
		} finally {
			accessControlService.unbanIp(clientIp)
		}
	}

	@Test
	@DisplayName("default rate limit을 초과하면 429를 반환한다")
	fun rateLimitExceededReturns429() {
		val clientIp = nextTestIp()

		repeat(3) {
			mockMvcTester
				.get()
				.uri(PROTECTED_PATH)
				.header("Authorization", "Bearer $accessToken")
				.header("X-Forwarded-For", clientIp)
				.exchange()
				.assertThat()
				.hasStatusOk()
		}

		val limited =
			mockMvcTester
				.get()
				.uri(PROTECTED_PATH)
				.header("Authorization", "Bearer $accessToken")
				.header("X-Forwarded-For", clientIp)
				.exchange()

		limited.assertThat().hasStatus(HttpStatus.TOO_MANY_REQUESTS)
		assertThat(limited.response.getHeader("Retry-After")).isNotBlank()
		assertThat(limited.response.getHeader("X-RateLimit-Remaining")).isEqualTo("0")
	}

	@Test
	@DisplayName("관리 API는 인증 없이 호출할 수 없다")
	fun managementApiRequiresAuthentication() {
		val targetIp = nextTestIp()
		val banBody =
			mapper.writeValueAsString(
				mapOf(
					"ip" to targetIp,
					"ttlSeconds" to 60,
					"reason" to "auth-check"
				)
			)

		// admin SecurityConfig는 AuthenticationEntryPoint 미설정 → 무토큰 시 403 Access Denied
		mockMvcTester
			.post()
			.uri(BAN_API)
			.contentType(MediaType.APPLICATION_JSON)
			.content(banBody)
			.exchange()
			.assertThat()
			.hasStatus(HttpStatus.FORBIDDEN)

		mockMvcTester
			.get()
			.uri(BAN_API)
			.exchange()
			.assertThat()
			.hasStatus(HttpStatus.FORBIDDEN)

		mockMvcTester
			.delete()
			.uri("$BAN_API?ip=$targetIp")
			.exchange()
			.assertThat()
			.hasStatus(HttpStatus.FORBIDDEN)
	}

	@Test
	@DisplayName("banned IP도 관리 API unban으로 탈출할 수 있다")
	fun bannedIpCanEscapeViaManagementUnban() {
		val escapeIp = nextTestIp()
		val banBody =
			mapper.writeValueAsString(
				mapOf(
					"ip" to escapeIp,
					"ttlSeconds" to 600,
					"reason" to "escape-path"
				)
			)

		// given: 관리 API로 ban
		mockMvcTester
			.post()
			.uri(BAN_API)
			.header("Authorization", "Bearer $accessToken")
			.contentType(MediaType.APPLICATION_JSON)
			.content(banBody)
			.exchange()
			.assertThat()
			.hasStatusOk()

		try {
			// when: ban된 IP로 일반 API 호출 → 403
			mockMvcTester
				.get()
				.uri(PROTECTED_PATH)
				.header("Authorization", "Bearer $accessToken")
				.header("X-Forwarded-For", escapeIp)
				.exchange()
				.assertThat()
				.hasStatus(HttpStatus.FORBIDDEN)

			// when: 동일 banned IP로 unban 관리 API 호출 → 통과 (management path)
			mockMvcTester
				.delete()
				.uri("$BAN_API?ip=$escapeIp")
				.header("Authorization", "Bearer $accessToken")
				.header("X-Forwarded-For", escapeIp)
				.exchange()
				.assertThat()
				.hasStatusOk()

			// then: unban 후 일반 API 재개
			mockMvcTester
				.get()
				.uri(PROTECTED_PATH)
				.header("Authorization", "Bearer $accessToken")
				.header("X-Forwarded-For", escapeIp)
				.exchange()
				.assertThat()
				.hasStatusOk()
		} finally {
			accessControlService.unbanIp(escapeIp)
		}
	}

	/**
	 * RFC 2544 벤치마크 대역(198.18.0.0/15)에서 실행마다 고유 IP를 만든다.
	 * Redis rate-limit 키 충돌을 피하기 위해 nanoTime·UUID·시퀀스를 조합한다.
	 */
	private fun nextTestIp(): String {
		val uuid = UUID.randomUUID()
		val seq = IP_SEQ.incrementAndGet()
		val b = Math.floorMod((uuid.mostSignificantBits xor System.nanoTime()).toInt(), 2) + 18
		val c = Math.floorMod((uuid.leastSignificantBits xor seq.toLong()).toInt(), 254) + 1
		val d = Math.floorMod((uuid.hashCode() xor seq * 31), 254) + 1
		return "198.$b.$c.$d"
	}
}
